Dev input vignette (#1348)
NOAA-FIMS/FIMS: vignettes/fims-input-data.Rmd

---
title: "Input data for FIMS: more detail"
output: github_document
vignette: >

>   %\VignetteIndexEntry{Input data for FIMS: more detail}

Reviewer: ```suggestion
title: "Introducing FIMS input data"
output: github_document
vignette: >
  %\VignetteIndexEntry{Introducing FIMS input data}
```

  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Set local options so vignette output is clean and consistent
# (avoids color codes and subtle formatting in rendered docs)
withr::local_options(list(
  cli.num_colors = 1,
  crayon.enabled = FALSE,
  pillar.subtle = FALSE
))
```

## Other sources of information
This vignette is an expansion of the section on "Data" in the [FIMS overview vignette](https://noaa-fims.github.io/FIMS/articles/fims-demo.html). It provides more details about the data structure needed for a FIMS model and how to prepare your data for use in a FIMS model using the `FIMSFrame()` function.

> The FIMS cheatsheet (https://github.com/NOAA-FIMS/FIMS/blob/dev/pkgdown/assets/fims-cheatsheet.pdf) includes a quick summary of the input data format.

Reviewer: ```suggestion
The [FIMS cheatsheet](https://github.com/NOAA-FIMS/FIMS/blob/dev/pkgdown/assets/fims-cheatsheet.pdf) includes a quick summary of the input data format.
```

The help page for the `FIMS::data_big` data set (https://noaa-fims.github.io/FIMS/reference/data_big.html or via `?FIMS::data_big`) provides detailed information about the input data format.

Note: this vignette does not include anything about model configuration, which is covered in the overview vignette linked above.

## Data format

> The input data for a FIMS model is a single long data frame with the following columns: `type`, `name`, `age`, `length`, `timing`, `value`, and `unit`. Both the single table and long format are a contrast to the input format for some legacy stock assessment models where the data are provided in multiple tables, each with a different format. The SS3 data file, for instance, has separate tables for catch, indices, age compositions, length compositions, with wide tables for the composition data (each age or length bin has a different column). The use of a single, long table for data input used by FIMS makes it easier to summarize and filter the data across data types (e.g. by fleet, time step, age, etc.), as well as being better suited to the widely-used tidyverse collection of packages in R.

Reviewer: ```suggestion
The input data for a FIMS model is a single long data frame with the following columns: `r glue::glue_collapse(colnames(data_big), sep = ", ", last = ", and ")`. Both the single table and long format are in contrast to the input format for some legacy stock assessment models where the data are provided in multiple tables, each with a different format. The SS3 data file, for instance, has separate tables for catch, indices, age compositions, length compositions, with wide tables for the composition data (each age or length bin has a different column). The use of a single, long table for data input used by FIMS makes it easier to summarize and filter the data across data types (e.g., by fleet, time step, age, etc.), as well as being better suited to the widely-used tidyverse collection of packages in R.
```

```{r data_big-landings}
FIMS::data_big |>
  dplyr::filter(type == "landings") |>
  head(3)
```

```{r data_big-landings-plot}
#| fig.alt: >
#|   Barplot shows an increase in landings over the timing values 1 to 8, then noisy but stable values until the final bar at timing equals 30.

library(ggplot2)
FIMS::data_big |>
  dplyr::filter(type == "landings") |>
  ggplot(aes(x = timing, y = value)) +
  geom_bar(stat = "identity")
```

### Indices (`type == "index"`)

> The index data type contains the relative or absolute abundance index for each time step. The inputs are similar to landings, in having `NA` values for age and length, `mt` for `unit`, and index uncertainty as the standard deviation of the logged observations. A `log_q` parameter can be fixed or estimated for each index.

Reviewer: Should we have information about q here because we are only talking about data? Also, you can fit indices and landings using numbers we just don't have an example of doing it. And, I don't think that the wrapper functions are set up to help you. Thus, it is only available for the power user.

```{r data_big-indices}
FIMS::data_big |>
  dplyr::filter(type == "index") |>
  head(3)
```

```{r data_big-indices-plot}
#| fig.alt: >
#|   Points with uncertainty shows a steadily decreasing index over time steps 1 to 30 with increasing precision for the later observations.
FIMS::data_big |>
  dplyr::filter(type == "index") |>
  ggplot(aes(x = timing, y = value)) +
  geom_point() +
  geom_pointrange(aes(
    ymin = qlnorm(p = 0.025, meanlog = log(value), sdlog = uncertainty),
    ymax = qlnorm(p = 0.975, meanlog = log(value), sdlog = uncertainty)
  )) +
  geom_hline(yintercept = 0)
```

### Age compositions (`type == "age_comp"`)

> The age composition data type contains the age distribution of a fleet for a time step. The `age` column contains the age, which needs to start at age 0, and the `value` column contains either the number or the proportion of the catch or survey in each age bin. The `unit` column should be either "number" or "proportion". If "proportion" is used, the values are checked at a later step to confirm that they sum to 1.0. The `uncertainty` column should be the input sample size for the time step (equal across all ages within that time step).

Reviewer: We might also want to note that uncertainty is multiplied by proportions or numbers to get the values that are used in the likelihood.

  ggplot2::ggplot(
    ggplot2::aes(x = age, y = value)
  ) +
  ggplot2::geom_bar(stat = "identity") +
  ggplot2::xlab("Age") +
  ggplot2::ylab("Number") +
  ggplot2::facet_wrap(~timing) +
  ggplot2::scale_x_continuous(breaks = 1:12)
```

### Length compositions (`type == "length_comp"`)
The length composition data are similar to the age compositions, except with values in the `length` column and `age = NA`. FIMS does not yet have parametric growth implemented, so the length bins can be any values as long as the distribution of each age among those bins is provided by the `age_to_length_conversion` data type described below.

```{r data_big-length_comp}
FIMS::data_big |>
  dplyr::filter(type == "length_comp") |>
  head(3)
```
```{r data_big-length_comp-plot}
#| fig.alt: >
#|   Barplot showing length comp by time step with a mode around 450 mm which declines and becomes more sharply peaked over time.

FIMS::data_big |>
  dplyr::filter(type == "length_comp") |>
  ggplot2::ggplot(
    ggplot2::aes(x = length, y = value)
  ) +
  ggplot2::geom_bar(stat = "identity") +
  ggplot2::xlab("Length") +
  ggplot2::ylab("Number") +
  ggplot2::facet_wrap(~timing)
```

> ### Weight-at-age data (`type == "weight_at_age"`)

Reviewer: ```suggestion
### Weight-at-age data (`type == "weight_at_age"`)

```

The weight-at-age values for the `data_big` are identical across time steps, and are provided for timing = 1 to 31 and associated with name = "fleet1".

```{r data_big-weight_at_age}
# first and last entries  of the weight-at-age data
FIMS::data_big |>
  dplyr::filter(type == "weight_at_age") |>
  head(3)
FIMS::data_big |>
  dplyr::filter(type == "weight_at_age") |>
  tail(3)

# check that the values are identical across time steps within each age
FIMS::data_big |>
  dplyr::filter(type == "weight_at_age") |>
  dplyr::group_by(age) |>
  dplyr::summarize(n_distinct = dplyr::n_distinct(value))
```

```{r data_big-weight_at_age-plot}
#| fig.alt: >
#|   Line plot showing weight at age showing identical monotonically increasing weight-at-age values ages that are the same across time steps.
FIMS::data_big |>
  dplyr::filter(type == "weight_at_age") |>
  ggplot2::ggplot(
    ggplot2::aes(x = age, y = value)
  ) +
  ggplot2::geom_line(color = "blue") +
  ggplot2::xlab("Age") +
  ggplot2::ylab("Weight (mt)") +
  ggplot2::facet_wrap(~timing) +
  ggplot2::scale_x_continuous(breaks = 1:12) +
  geom_hline(yintercept = 0)
```

> ### Age-to-length conversion (`type == "age_to_length_conversion"`)

Reviewer: ```suggestion
### Age-to-length conversion (`type == "age_to_length_conversion"`)

```
